### ISSUE TITLE
How do you get IEndpointInstance when using the Host?

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/NServiceBus.Host#85]
### Problem
Documentation indicates that  to use `IEndpointInstance` 
http://docs.particular.net/nservicebus/messaging/send-a-message#outside-a-message-handler

However, how does one get a hold of IEndpointInstance, especially since that it's not injected?

Relates to: https://github.com/Particular/V6Launch/issues/27

### Proposed Solution

- Get the IMessageSession by implementing IWantToRunWhenEndpointStartsAndStops. However this is not supported by all containers and is also not the best way. 

- [Proposal from @danielmarbach to change the `IConfigureThisEndpointInterface` to provide access]( https://github.com/Particular/NServiceBus.Host/issues/85#issuecomment-219771733). Will require changes to the host.

**Comment by indualagarsamy** [comment]
@Particular/nservicebus-maintainers - please review and comment, provide relevant code snippets or PR to doco.

**Comment by andreasohlund** [comment]
Why do you need the endpoint instance?

The idea is to use the IMessageSession instead => https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/StartableAndStoppable/IWantToRunWhenEndpointStartsAndStops.cs#L18

**Comment by danielmarbach** [comment]
@andreasohlund If you want to bind `IMessageSession` on the container you know have to do the following:

* Implement `IWantToRunWhenEndpointStartsAndStops`
* Inject the container registration API somehow into that implementation
* Register the `IMessageSession`

The biggest problem is since  `IWantToRunWhenEndpointStartsAndStops` is scanned and registered on the container, depending on the container it might be already closed for registrations when the implementation is resolved

**Comment by andreasohlund** [comment]
@indualagarsamy should we just add this to the doco?

* when self hosting you get the instance from bus.Start
* when using the azure/windows host you get it by implementing `IWantToRunWhenTheEndpointStarts`

**Comment by indualagarsamy** [comment]
@andreasohlund - How do you inject the IManageSession into the container at the IWantToRunWhenEndpointStartsAndStops. If you know the exact steps, can you detail them out here?

**Comment by andreasohlund** [comment]
That would be container specific. What container is the user on?

**Comment by danielmarbach** [comment]
@andreasohlund Let me repeat my point:

The implementation of IConfigureThisEndpoint will be the owner of the container. After we called `Customize` we will start the endpoint and that means we already resolved from the user container. Depending on the container that means the container is already locked. So you might not be able to register `IEndpointInstance`

**Comment by andreasohlund** [comment]
Agreed, but I know that StructureMap can handle this so we can use that if we want to include an example.

But really the best was is to pass the instance via method injection instead

**Comment by danielmarbach** [comment]
One way to resolve this problem could be to push a bit more responsibility towards the implementor of `IConfigureThisEndpoint`. I have a proposal, running it with @indualagarsamy

**Comment by danielmarbach** [comment]
@andreasohlund how about we change the interface to something like

```
    class MyEndpointConfig : IConfigureThisEndpoint
    {
        public async Task<IEndpointInstance> Customize(EndpointConfiguration configuration)
        {
            configuration.UseContainer<Ninject>();

            var endpointInstance = await Endpoint.Start(configuration);
            return endpointInstance;
        }
    }
```

we could even generate the last two lines with the template we ship. This would make the host much more aligned with custom hosting and moves the responsibility of creating the endpoint to the owner of all the information necessary to create the endpoint: The custom implementation of `IConfigureThisEndpoint`. 

Sample

```
    class MyEndpointConfig : IConfigureThisEndpoint
    {
        public async Task<IEndpointInstance> Customize(EndpointConfiguration configuration)
        {
            var kernel = new StandardKernel();
            configuration.UseContainer<Ninject>();

            var endpointInstance = await Endpoint.Start(configuration);
            kernel.Bind<IEndpointInstance>().ToConstant(endpointInstance);
            return endpointInstance;
        }
    }
```

**Comment by indualagarsamy** [comment]
Just had a discussion with @danielmarbach about this and I agree that this as a much better way than having to implement a IWantToRunWhenEndpointStartsAndStops just to get a hold of IManageSession, which is a completely different interface.

Holding on to IEndpointInstance is a much better way because for the user it ties in to the same model, i.e. one less thing to learn. Regardless of the hosting model, it's just one interface.

**Comment by andreasohlund** [comment]
What about calling `await endpointInstance.Stop()` when the windows service shuts down? 

Would that be on the user?

**Comment by bording** [comment]
Since this interface would now be responsbile for much more than just adding endpoint configuration, would it make sense to rename it as part of this change? Perhaps something more like `ICreateEndpoint`, with a `Create` method?

**Comment by danielmarbach** [comment]
The idea was the stop would still be done by the host. Only creation and
start is owned by the user implementation. But of course this is debatable
as well.

:+1: on renaming as well

**Comment by indualagarsamy** [comment]
@danielmarbach - I've thought about this after discussing on a different topic with @timbussmann. 
Here are the reasons why I think we should leave it as is and make it less accessible to get at the IEndpointInstance atleast when using the Host.

People that were previously using IBus (injected in their components), might try and replace it with IEndpointInstance as well. The scenarios where it would be problematic are:

If they're replacing an IBus in an custom component with IEndpointInstance, they could easily use the IEndpointInstance interface methods to send/publish instead of using the message handler context. Doing so will mean 
- that those messages that sent will not partake in the transaction scope,
- they won't be part of the outgoing batching operation of the message handler. 
- they won't have access to the same message headers such as correlation. 

By making it easy to obtain IEndpointInstance, then users could simply replace IBus with IEndpointInstance instead and we shouldn't make that path easy. 

I think we need documentation on this pronto! 

@Particular/nservicebus-maintainers - If you agree, then I will close the issue. If anything, if we can we should throw if messages are being sent using the wrong context. i.e. from inside the handler if messages are sent using IMessageSession instead of the IMessageHandlerContext. But that's a different issue.

**Comment by danielmarbach** [comment]
I agree on the danger and that we should not make it easy. In fact that was the initial idea from the start when @timbussmann and I started designed the handler context APIs. The problem is as it stand we make it impossible for users using the host in combination with a container which is locked after the first resolve to bind `IEndpointInstance`. Are we aware of the implications of that approach?

 It would mean all those users would need to switch to self-hosting (which might be desired). And here is the flaw we have, as I pointed out in self-hosting registering IEndpointInstance is already possible. So from that perspective not doing anything but doco we would not solve much.

**Comment by indualagarsamy** [comment]
@danielmarbach - See this PR. @SeanFeldman and I have been looking into this https://github.com/indualagarsamy/MigrateIBus/pull/2.
Do you see any downsides to that approach?

Switching to self-host is not the only option. See above. Do you still feel strongly about this? I believe if we have a working solution, we don't have to put it any more effort on the host.

**Comment by SeanFeldman** [comment]
> Do you see any downsides to that approach?

I want to highlight that that specific approach is _solely_ intended for the scenarios where component is not invoked by a handler (i.e. component is not within the handler context).

**Comment by indualagarsamy** [comment]
@SeanFeldman - I don't quite follow what you're saying as a downside 
What I am saying is that we don't need to go to extra measures to expose IEndpointInstance from the host. For people that use the host for the scenario where their component needs it, they can use the IMessageSession from the IWant.. interface. Do you see any problems with that?

**Comment by SeanFeldman** [comment]
We're in agreement @indualagarsamy

**Comment by indualagarsamy** [comment]
@andreasohlund @danielmarbach - if you guys agree, I can close this issue.

**Comment by danielmarbach** [comment]
@indualagarsamy I still feel strongly about it :D I'll raise a separate issue to bring together all the facts and gather a TF to solve this once and for all. In summary:

The sample PR you provided relies on the fact that the container allows mutation after the first resolve. In my point of view we should not encourage this since it is not only an anti-pattern but also adds another point to the list which hinders users to use mechanics like AutoFac container building and forces them to [updated existing containers](http://stackoverflow.com/questions/4998870/adding-services-after-container-has-been-built). Furthermore, it is another point which makes it hard to use SimpleInjector.

see https://simpleinjector.readthedocs.io/en/latest/pipeline.html and https://simpleinjector.readthedocs.io/en/latest/decisions.html#container-is-locked

// cc @dotnetjunkie

**Comment by danielmarbach** [comment]
@indualagarsamy opened up a new issue https://github.com/Particular/PlatformDevelopment/issues/819

**Comment by danielmarbach** [comment]
Closing this one in favour of https://github.com/Particular/PlatformDevelopment/issues/819

**Comment by dotnetjunkie** [comment]
@danielmarbach  is right. Changing the container after it has been resolved from not only complicates your application, but can easily lead to unexpected behavior; you will see this with any container that allows you to change a used container.

**Comment by indualagarsamy** [comment]
@dotnetjunkie @danielmarbach - good points. 

So we're back to square one. The only option when using NServiceBus.Host at the moment is to float the handler message context (which is the preferred way) or switch to self hosting.

**Comment by danielmarbach** [comment]
Not entirely. My proposal for IConfigureThisEndpoint would work. I will
sketch it out a bit more

**Comment by indualagarsamy** [comment]
@danielmarbach - Square one, meaning status quo. You have a proposal, but that needs to be fleshed out and all the subsequent questions answered, i.e. on the endpoint start and stop. Let's discuss this one  in plat dev and then document our findings / way forward here.

**Comment by indualagarsamy** [comment]
@danielmarbach - I am reopening this issue. Whatever is decided based on the PlatformDev meetings, please update this issue, if we are going to be making code changes to the host to support this.

**Comment by andreasohlund** [comment]
> reopened this issue because getting IMessageSession from IWantToRunWhenEndpointStartsAndStops is not really possible for all containers and is not preferred.

So we're saying that pushing users towards using method injection when using the host or switching to selfhosting is not an option?

**Comment by andreasohlund** [comment]
@indualagarsamy what should we do here, add guidance how to pass the message session via method injection as the prefered way? (and switch to selfhost if that's not possible)

**Comment by indualagarsamy** [comment]
@andreasohlund - Can you see if the guidance we have here is good? I worked on this a while ago. 
http://docs.particular.net/nservicebus/upgrades/5to6-moving-away-from-ibus#dependency-injection

Here's what I think:
- In a web scenario, hanging to the IMessageSession might be needed as it might be needed in controllers to send messages. But that's already covered in a self hosting scenarios. 
- In a NServiceBus.Host scenario, other than sending messages on startup, I can't come up with scenarios where a IMessageSession would be needed for later purposes that warrants it to be injected in the container for later purposes.
If you think the guidance is good enough or if you can further improve what we have, I am in favor of closing this issue.

**Comment by andreasohlund** [comment]
http://docs.particular.net/nservicebus/upgrades/5to6-moving-away-from-ibus#dependency-injection looks good to me.

Closing this then since https://github.com/Particular/NServiceBus/issues/3749 tracks what's left to do in terms of detecting misuse of the the message session

**Comment by sanilpaul** [comment]
@indualagarsamy I know this is closed. However, I couldn't find what am looking for anywhere in the docs. how would you recommend getting hold of IEndpointInstance/IMessageSession to do something like that's described in https://lostechies.com/jimmybogard/2012/08/13/reliable-job-scheduling-with-nservicebus-and-quartz-net/ within the context of an endpoint and using Nservicebus.Host. 

https://github.com/jbogard/QuartzNServiceBusSample/blob/master/QuartzNServiceBusSample/DoSomethingSchedule.cs

Basically I want to get IMessageSession in the IJob implementation. The trigger is the quartz.net and it works in an endpoint. but not in the context of a handler. We implemented quartz scheduler (https://github.com/jbogard/QuartzNServiceBusSample/blob/master/QuartzNServiceBusSample/QuartzService.cs) as a NSB feature with a startup task that starts up the scheduler. 
Any help would be greatly appreciated. We are trying to upgrade our solution from NSB5 to NSB6.

**Comment by sanilpaul** [comment]
@ramonsmits Any updates?

**Comment by indualagarsamy** [comment]
@Particular/host-maintainers - Can you comment on @sanilpaul's comments above?

**Comment by ramonsmits** [comment]
@sanilpaul I'm working on a POC with the latest version of quartz.net

**Comment by danielmarbach** [comment]
@sanilpaul

The feature startup task gets access to the message session. You pass that session into the job factory which creates the job implementations. You then set the jobfactory on the scheduler before you start it.

**Comment by sanilpaul** [comment]
@ramonsmits @danielmarbach Thank you for the recommendation.  That's how we ended up doing and is working fine. I had a support ticket(00026956) for the same.

**Comment by ramonsmits** [comment]
@sanilpaul Thanks for mentioning the ticket number. I looked at your solution and wanted to note that it uses `InstancePerUnitOfWork`. As in that context there is no unit of work this means that the job instance is only created once, basically having `SingleInstance`.  If you are only sending a message in your jobs then that is fine as basically the job is stateless and can be reused but if you let it inject other resources then this might require `InstancePerCall` to have a new job instance created each time the job is triggered.

**Comment by sanilpaul** [comment]
@ramonsmits Thank you for looking into the solution. That pointer is very helpful.
